### DISCUSSION TITLE
[WIP] Support Ray monitoring

### DISCUSSION THREAD

**Discussion by Superskyyy** [apache/skywalking#11905] [Ideas]
Ray is a popular distributed computing framework that powered the training of all kinds of AI workloads, e.g., ChatGPT. My current work goes deep into the Ray side so I'd like to start working on monitoring Ray with SkyWalking, I will draft a SWIP first and share here.

**Comment by wu-sheng** [comment]
Is that a Python framework?

**Reply by Superskyyy** [reply]
It's core is c++ but upper layer is Python. They already have some metrics and traces exposed there https://docs.ray.io/en/latest/ray-observability/key-concepts.html#metrics

**Reply by wu-sheng** [reply]
I am highly doubt the expose trace works. Because if you need native traces to analysis, general OTEL/Zipkin traces don't work.

**Reply by Superskyyy** [reply]
> I am highly doubt the expose trace works. Because if you need native traces to analysis, general OTEL/Zipkin traces don't work.

Yeah I will just write some SkyWalking python plugin for application layer tracing. The Otel trace doesn't even properly function last time I saw discussions around it.

**Reply by wu-sheng** [reply]
Meanwhile, if this is a client framework, why SWIP is considered? Is there a backend relative change?

**Reply by Superskyyy** [reply]
Both client and server sides are needed since the Ray cluster performance is also critical. Ray backend already ships metrics for scraping, so I will add a dedicated dashboard.

**Reply by wu-sheng** [reply]
That makes sense. Please be advised, current service hierarchy concept, which requires you to build the link between virtual services and server services. I am not sure which kind of virtual service you could treat the remote from the client side.

**Reply by Superskyyy** [reply]
Ok, I need to read about that concept to understand a bit more.

**Reply by wu-sheng** [reply]
We have built this into v10, https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/service-hierarchy/

Also, the showcase has some of that hierarchy recognition already.
